pkg/controller/user/user_controller.go

```diff
@@ -239,6 +243,48 @@ func (c *Controller) reconcile(key string) error {
 		return err
 	}
 
+	// name of your custom finalizer
+	finalizer := "finalizers.kubesphere.io/users"
+
+	if user.ObjectMeta.DeletionTimestamp.IsZero() {
+		// The object is not being deleted, so if it does not have our finalizer,
+		// then lets add the finalizer and update the object.
+		if !sliceutil.HasString(user.Finalizers, finalizer) {
+			user.ObjectMeta.Finalizers = append(user.ObjectMeta.Finalizers, finalizer)
+
+			if user, err = c.ksClient.IamV1alpha2().Users().Update(user); err != nil {
+				return err
+			}
+		}
+	} else {
+		// The object is being deleted
+		if sliceutil.HasString(user.ObjectMeta.Finalizers, finalizer) {
+
+			klog.V(4).Infof("delete user %s", key)
+			if err = c.ldapClient.Delete(key); err != nil && err != ldapclient.ErrUserNotExists {
+				klog.Error(err)
+				return err
+			}
+
+			// remove our finalizer from the list and update it.
+			user.Finalizers = sliceutil.RemoveString(user.ObjectMeta.Finalizers, func(item string) bool {
+				return item == finalizer
+			})
+
+			if _, err := c.ksClient.IamV1alpha2().Users().Update(user); err != nil {
+				return err
+			}
+		}
+
+		// Our finalizer has finished, so the reconciler can do nothing.
+		return nil
+	}
+
+	if err = c.ldapSync(user); err != nil {
```

Comment:
What if ldapClient is nil ?

Reply:
Is any comments on this?

Reply:
LDAP needs to be enabled during installation, must not be nil. Use simple ldap client in debug mode.

Reply:
Then we'd better check client during controller initialization, make sure we don't get a nil ldap client

Reply:
Is it necessary to use simple LDAP client if LDAP options not provide?

Reply:
No, if ldap not provided, just return error. Simple Ldap is only used for testing

Reply:
done